File: app/src/main/java/com/digicoffer/lauditor/CommonFiles/CommonAdapters/CommonMultiSelectionAdapter.java
package com.digicoffer.lauditor.CommonFiles.CommonAdapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.digicoffer.lauditor.R;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class CommonMultiSelectionAdapter extends RecyclerView.Adapter<CommonMultiSelectionAdapter.ViewHolder> {

    private Context context;
    private List<String> allPracticeAreas;
    private List<String> filteredPracticeAreas;
    private Set<String> selectedPracticeAreas;
    private Set<String> suggestedItems = new HashSet<>();  // ← ADD THIS
    private boolean showSuggestIcon = false;               // ← ADD THIS
    private OnSelectionChangeListener listener;

    public interface OnSelectionChangeListener {
        void onSelectionChanged(List<String> selectedItems);
    }

    public CommonMultiSelectionAdapter(Context context, List<String> practiceAreas,
                                       List<String> preSelectedItems,
                                       OnSelectionChangeListener listener) {
        this.context = context;
        this.allPracticeAreas = new ArrayList<>(practiceAreas);
        this.filteredPracticeAreas = new ArrayList<>(practiceAreas);
        this.selectedPracticeAreas = new HashSet<>(preSelectedItems != null ? preSelectedItems : new ArrayList<>());
        this.listener = listener;
    }
    public void setSuggestedItems(List<String> suggested, boolean showIcon) {
        this.suggestedItems = new HashSet<>(suggested != null ? suggested : new ArrayList<>());
        this.showSuggestIcon = showIcon;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.select_multiple_checkbox, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        String practiceArea = filteredPracticeAreas.get(position);

        holder.tvPracticeName.setText(practiceArea);
        boolean isOtherOption = practiceArea.startsWith("+ Other");

        if (isOtherOption) {
            holder.checkBox.setVisibility(View.GONE);
            holder.checkBox.setChecked(false);
        } else {
            holder.checkBox.setVisibility(View.VISIBLE);
            boolean isSelected = selectedPracticeAreas.contains(practiceArea);
            holder.checkBox.setChecked(isSelected);
            holder.checkBox.setEnabled(true);
            holder.checkBox.setAlpha(1f);
        }

        // ── iv_suggest: only visible for services + only for suggested items ──
        if (holder.ivSuggest != null) {
            boolean isSuggested = showSuggestIcon && suggestedItems.contains(practiceArea);
            holder.ivSuggest.setVisibility(isSuggested ? View.VISIBLE : View.GONE);
        }

        holder.checkBox.setOnClickListener(v -> {
            if (holder.checkBox.isChecked()) {
                selectedPracticeAreas.add(practiceArea);
                if (listener != null) listener.onSelectionChanged(new ArrayList<>(selectedPracticeAreas));
            } else {
                selectedPracticeAreas.remove(practiceArea);
                if (listener != null) listener.onSelectionChanged(new ArrayList<>(selectedPracticeAreas));
            }
        });

        holder.itemView.setOnClickListener(v -> holder.checkBox.performClick());
    }

    public void unselectItem(String itemName) {
        if (itemName == null || itemName.trim().isEmpty()) return;

        selectedPracticeAreas.remove(itemName);

        for (int i = 0; i < filteredPracticeAreas.size(); i++) {
            if (filteredPracticeAreas.get(i).equals(itemName)) {
                notifyItemChanged(i);
                break;
            }
        }

        if (listener != null) {
            listener.onSelectionChanged(new ArrayList<>(selectedPracticeAreas));
        }
    }

    public List<String> getAllItems() {
        return new ArrayList<>(allPracticeAreas); // return a copy of the backing list
    }

    public void selectItem(String itemName) {
        if (itemName == null || itemName.trim().isEmpty()) return;

        selectedPracticeAreas.add(itemName);

        for (int i = 0; i < filteredPracticeAreas.size(); i++) {
            if (filteredPracticeAreas.get(i).equals(itemName)) {
                notifyItemChanged(i);
                break;
            }
        }

        if (listener != null) {
            listener.onSelectionChanged(new ArrayList<>(selectedPracticeAreas));
        }
    }

    // NEW METHOD: Update the entire data set
    public void updateData(List<String> newData) {
        if (newData == null) {
            newData = new ArrayList<>();
        }
        this.allPracticeAreas = new ArrayList<>(newData);
        this.filteredPracticeAreas = new ArrayList<>(newData);
        notifyDataSetChanged();
    }

    // NEW METHOD: Set which items are selected
    public void setSelectedItems(List<String> selectedItems) {
        if (selectedItems == null) {
            selectedItems = new ArrayList<>();
        }
        this.selectedPracticeAreas = new HashSet<>(selectedItems);
        notifyDataSetChanged();

        // Notify listener about the selection change
        if (listener != null) {
            listener.onSelectionChanged(new ArrayList<>(selectedPracticeAreas));
        }
    }

    // NEW METHOD: Add a single selected item
    public void addSelectedItem(String item) {
        if (item == null || item.isEmpty()) return;
        selectedPracticeAreas.add(item);
        notifyDataSetChanged();

        if (listener != null) {
            listener.onSelectionChanged(new ArrayList<>(selectedPracticeAreas));
        }
    }

    // NEW METHOD: Remove a single selected item
    public void removeSelectedItem(String item) {
        if (item == null || item.isEmpty()) return;
        selectedPracticeAreas.remove(item);
        notifyDataSetChanged();

        if (listener != null) {
            listener.onSelectionChanged(new ArrayList<>(selectedPracticeAreas));
        }
    }

    // NEW METHOD: Clear all selected items
    public void clearAllSelected() {
        selectedPracticeAreas.clear();
        notifyDataSetChanged();

        if (listener != null) {
            listener.onSelectionChanged(new ArrayList<>());
        }
    }

    @Override
    public int getItemCount() {
        return filteredPracticeAreas.size();
    }

    public void filter(String query) {
        filteredPracticeAreas.clear();

        if (query == null || query.trim().isEmpty()) {
            filteredPracticeAreas.addAll(allPracticeAreas);
        } else {
            String lowerCaseQuery = query.toLowerCase().trim();
            for (String item : allPracticeAreas) {
                if (item.toLowerCase().contains(lowerCaseQuery)) {
                    filteredPracticeAreas.add(item);
                }
            }
        }

        notifyDataSetChanged();
    }

    public List<String> getSelectedItems() {
        return new ArrayList<>(selectedPracticeAreas);
    }

    public int getSelectedCount() {
        return selectedPracticeAreas.size();
    }

    public void clearSelection() {
        selectedPracticeAreas.clear();
        notifyDataSetChanged();
        if (listener != null) {
            listener.onSelectionChanged(new ArrayList<>());
        }
    }

    public void renameItem(String oldName, String newName) {
        if (oldName == null || newName == null) return;
        if (oldName.trim().isEmpty() || newName.trim().isEmpty()) return;

        // Update in allPracticeAreas
        int allIndex = allPracticeAreas.indexOf(oldName);
        if (allIndex != -1) {
            allPracticeAreas.set(allIndex, newName);
        }

        // Update in filteredPracticeAreas
        int filteredIndex = filteredPracticeAreas.indexOf(oldName);
        if (filteredIndex != -1) {
            filteredPracticeAreas.set(filteredIndex, newName);
            notifyItemChanged(filteredIndex);
        }

        // Update in selectedPracticeAreas if it was selected
        if (selectedPracticeAreas.contains(oldName)) {
            selectedPracticeAreas.remove(oldName);
            selectedPracticeAreas.add(newName);
            if (listener != null) {
                listener.onSelectionChanged(new ArrayList<>(selectedPracticeAreas));
            }
        }
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        TextView tvPracticeName;
        CheckBox checkBox;
        ImageView ivSuggest;   // ← ADD THIS

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            tvPracticeName = itemView.findViewById(R.id.tv_tm_name);
            checkBox = itemView.findViewById(R.id.chk_selected);
            ivSuggest = itemView.findViewById(R.id.iv_suggest);  // ← ADD THIS
        }
    }
}

File: app/src/main/java/com/digicoffer/lauditor/CommonFiles/CommonAdapters/CommonMultiSelectionAdapter.kt
package com.digicoffer.lauditor.CommonFiles.CommonAdapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.digicoffer.lauditor.R
import java.util.ArrayList
import java.util.HashSet

class CommonMultiSelectionAdapter(
    private val context: Context,
    practiceAreas: List<String>?,
    preSelectedItems: List<String>?,
    private val listener: OnSelectionChangeListener?
) : RecyclerView.Adapter<CommonMultiSelectionAdapter.ViewHolder>() {

    private var allPracticeAreas: ArrayList<String> = if (practiceAreas != null) ArrayList(practiceAreas) else ArrayList()
    private var filteredPracticeAreas: ArrayList<String> = if (practiceAreas != null) ArrayList(practiceAreas) else ArrayList()
    private var selectedPracticeAreas: HashSet<String> = if (preSelectedItems != null) HashSet(preSelectedItems) else HashSet()
    private var suggestedItems: HashSet<String> = HashSet()
    private var showSuggestIcon = false

    interface OnSelectionChangeListener {
        fun onSelectionChanged(selectedItems: List<String>)
    }

    fun setSuggestedItems(suggested: List<String>?, showIcon: Boolean) {
        this.suggestedItems = if (suggested != null) HashSet(suggested) else HashSet()
        this.showSuggestIcon = showIcon
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.select_multiple_checkbox, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val practiceArea = filteredPracticeAreas[position]

        holder.tvPracticeName.text = practiceArea
        val isOtherOption = practiceArea.startsWith("+ Other")

        if (isOtherOption) {
            holder.checkBox.visibility = View.GONE
            holder.checkBox.isChecked = false
        } else {
            holder.checkBox.visibility = View.VISIBLE
            val isSelected = selectedPracticeAreas.contains(practiceArea)
            holder.checkBox.isChecked = isSelected
            holder.checkBox.isEnabled = true
            holder.checkBox.alpha = 1f
        }

        if (holder.ivSuggest != null) {
            val isSuggested = showSuggestIcon && suggestedItems.contains(practiceArea)
            holder.ivSuggest.visibility = if (isSuggested) View.VISIBLE else View.GONE
        }

        holder.checkBox.setOnClickListener {
            if (holder.checkBox.isChecked) {
                selectedPracticeAreas.add(practiceArea)
            } else {
                selectedPracticeAreas.remove(practiceArea)
            }
            listener?.onSelectionChanged(ArrayList(selectedPracticeAreas))
        }

        holder.itemView.setOnClickListener { holder.checkBox.performClick() }
    }

    fun unselectItem(itemName: String?) {
        if (itemName.isNullOrBlank()) return
        selectedPracticeAreas.remove(itemName)
        for (i in filteredPracticeAreas.indices) {
            if (filteredPracticeAreas[i] == itemName) {
                notifyItemChanged(i)
                break
            }
        }
        listener?.onSelectionChanged(ArrayList(selectedPracticeAreas))
    }

    fun getAllItems(): List<String> {
        return ArrayList(allPracticeAreas)
    }

    fun selectItem(itemName: String?) {
        if (itemName.isNullOrBlank()) return
        selectedPracticeAreas.add(itemName)
        for (i in filteredPracticeAreas.indices) {
            if (filteredPracticeAreas[i] == itemName) {
                notifyItemChanged(i)
                break
            }
        }
        listener?.onSelectionChanged(ArrayList(selectedPracticeAreas))
    }

    fun updateData(newData: List<String>?) {
        val data = newData ?: ArrayList()
        this.allPracticeAreas = ArrayList(data)
        this.filteredPracticeAreas = ArrayList(data)
        notifyDataSetChanged()
    }

    fun setSelectedItems(selectedItems: List<String>?) {
        val items = selectedItems ?: ArrayList()
        this.selectedPracticeAreas = HashSet(items)
        notifyDataSetChanged()
        listener?.onSelectionChanged(ArrayList(selectedPracticeAreas))
    }

    fun addSelectedItem(item: String?) {
        if (item.isNullOrEmpty()) return
        selectedPracticeAreas.add(item)
        notifyDataSetChanged()
        listener?.onSelectionChanged(ArrayList(selectedPracticeAreas))
    }

    fun removeSelectedItem(item: String?) {
        if (item.isNullOrEmpty()) return
        selectedPracticeAreas.remove(item)
        notifyDataSetChanged()
        listener?.onSelectionChanged(ArrayList(selectedPracticeAreas))
    }

    fun clearAllSelected() {
        selectedPracticeAreas.clear()
        notifyDataSetChanged()
        listener?.onSelectionChanged(ArrayList())
    }

    override fun getItemCount(): Int {
        return filteredPracticeAreas.size
    }

    fun filter(query: String?) {
        filteredPracticeAreas.clear()
        if (query.isNullOrBlank()) {
            filteredPracticeAreas.addAll(allPracticeAreas)
        } else {
            val lowerCaseQuery = query.lowercase().trim()
            for (item in allPracticeAreas) {
                if (item.lowercase().contains(lowerCaseQuery)) {
                    filteredPracticeAreas.add(item)
                }
            }
        }
        notifyDataSetChanged()
    }

    fun getSelectedItems(): List<String> {
        return ArrayList(selectedPracticeAreas)
    }

    fun getSelectedCount(): Int {
        return selectedPracticeAreas.size
    }

    fun clearSelection() {
        selectedPracticeAreas.clear()
        notifyDataSetChanged()
        listener?.onSelectionChanged(ArrayList())
    }

    fun renameItem(oldName: String?, newName: String?) {
        if (oldName.isNullOrBlank() || newName.isNullOrBlank()) return

        val allIndex = allPracticeAreas.indexOf(oldName)
        if (allIndex != -1) {
            allPracticeAreas[allIndex] = newName
        }

        val filteredIndex = filteredPracticeAreas.indexOf(oldName)
        if (filteredIndex != -1) {
            filteredPracticeAreas[filteredIndex] = newName
            notifyItemChanged(filteredIndex)
        }

        if (selectedPracticeAreas.contains(oldName)) {
            selectedPracticeAreas.remove(oldName)
            selectedPracticeAreas.add(newName)
            listener?.onSelectionChanged(ArrayList(selectedPracticeAreas))
        }
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val tvPracticeName: TextView = itemView.findViewById(R.id.tv_tm_name)
        val checkBox: CheckBox = itemView.findViewById(R.id.chk_selected)
        val ivSuggest: ImageView? = itemView.findViewById(R.id.iv_suggest)
    }
}
